pkg/costattribution/sample_tracker.go

```diff
@@ -0,0 +1,305 @@
+	overflowLabels[len(trackedLabels)+1] = overflowValue
+
+	tracker := &SampleTracker{
+		userID:           userID,
+		labels:           trackedLabels,
+		maxCardinality:   limit,
+		observed:         make(map[string]*observation),
+		cooldownDuration: cooldown,
+		logger:           logger,
+		overflowLabels:   overflowLabels,
+		overflowCounter:  observation{},
+	}
+
+	variableLabels := slices.Clone(trackedLabels)
+	variableLabels = append(variableLabels, tenantLabel, "reason")
+	tracker.discardedSampleAttribution = prometheus.NewDesc("cortex_discarded_attributed_samples_total",
+		"The total number of samples that were discarded per attribution.",
+		variableLabels,
+		prometheus.Labels{trackerLabel: defaultTrackerName})
+
+	tracker.receivedSamplesAttribution = prometheus.NewDesc("cortex_distributor_received_attributed_samples_total",
+		"The total number of samples that were received per attribution.",
+		variableLabels[:len(variableLabels)-1],
+		prometheus.Labels{trackerLabel: defaultTrackerName})
+	return tracker
+}
+
+func (st *SampleTracker) hasSameLabels(labels []string) bool {
+	return slices.Equal(st.labels, labels)
+}
+
+var bufferPool = sync.Pool{
+	New: func() interface{} {
+		return new(bytes.Buffer)
+	},
+}
+
+func (st *SampleTracker) cleanupTrackerAttribution(key string) {
+	st.observedMtx.Lock()
+	defer st.observedMtx.Unlock()
+	delete(st.observed, key)
+}
+
+func (st *SampleTracker) Collect(out chan<- prometheus.Metric) {
+	if st.overflowSince.Load() > 0 {
+		out <- prometheus.MustNewConstMetric(st.receivedSamplesAttribution, prometheus.CounterValue, st.overflowCounter.receivedSample.Load(), st.overflowLabels[:len(st.overflowLabels)-1]...)
+		out <- prometheus.MustNewConstMetric(st.discardedSampleAttribution, prometheus.CounterValue, st.overflowCounter.totalDiscarded.Load(), st.overflowLabels...)
+		return
+	}
+	// We don't know the performance of out receiver, so we don't want to hold the lock for too long
+	var prometheusMetrics []prometheus.Metric
+	st.observedMtx.RLock()
+	for key, o := range st.observed {
+		keys := strings.Split(key, string(sep))
+		keys = append(keys, st.userID)
+		if o.receivedSample.Load() > 0 {
+			prometheusMetrics = append(prometheusMetrics, prometheus.MustNewConstMetric(st.receivedSamplesAttribution, prometheus.CounterValue, o.receivedSample.Load(), keys...))
+		}
+		o.discardedSampleMtx.Lock()
+		for reason, discarded := range o.discardedSample {
+			prometheusMetrics = append(prometheusMetrics, prometheus.MustNewConstMetric(st.discardedSampleAttribution, prometheus.CounterValue, discarded.Load(), append(keys, reason)...))
+		}
+		o.discardedSampleMtx.Unlock()
+	}
+	st.observedMtx.RUnlock()
+
+	for _, m := range prometheusMetrics {
+		out <- m
+	}
+}
+
+func (st *SampleTracker) IncrementDiscardedSamples(lbs []mimirpb.LabelAdapter, value float64, reason string, now time.Time) {
+	if st == nil {
+		return
+	}
+	st.updateCountersWithLabelAdapter(lbs, now, 0, value, &reason)
```

Comment:
Nit: I don't think this method call adds value: that method is 6 lines long and it's not called anywhere else. Just put its contents here.

Reply:
removed here 73a988160668f510d1b2d89866427fc3fe47d5c8